### PR TITLE
Store lyrics in the database as media streams

### PR DESCRIPTION
**Pull request by crobibero**
**Changes**

Store the external lyric files as media streams like subtitles.

**Issues**

Working plugin: https://github.com/crobibero/jellyfin-plugin-lrclib

### CODE REVIEW

File: MediaBrowser.Providers/Lyric/ILyricProvider.cs
Comment (nielsvanvelzen) [inline MediaBrowser.Providers/Lyric/ILyricProvider.cs:29]: This will break the plugin support I added in #9920. With BaseItem a plugin can find lyrics based on metadata (artist/track name) from various sources, including the file system but also online services etc.

Comment (crobibero) [reply MediaBrowser.Providers/Lyric/ILyricProvider.cs]: Good call, I've reverted that change and now plugins can specify whether the item has lyrics and can also provide them

**Comment by nielsvanvelzen** [comment]
I think it would make more sense to have the LyricManager store the results in the database and not make the providers responsible for that. With the current approach a cloud-provider would still hold up item responses.

**Comment by DomiStyle** [comment]
What's the preferred way to deal with embedded lyrics? They are currently not stored in database either.

They are provided by TagLib (`tags.Lyrics`) though, should they be added as media streams as well if available?

**Review by JPVenson — APPROVED** [review]

**Review by Gauvino — APPROVED** [review]

**Comment by NicholasFlamy** [comment]
Thanks for your work. Can't wait! 👏👏👏

**Review by Shadowghost — APPROVED** [review]

File: Jellyfin.Api/Models/LyricDtos/UploadLyricDto.cs
Comment (nielsvanvelzen) [inline Jellyfin.Api/Models/LyricDtos/UploadLyricDto.cs:20]: I think this isn't used

File: Jellyfin.Api/Controllers/LyricController.cs
Comment (nielsvanvelzen) [inline Jellyfin.Api/Controllers/LyricController.cs:243]:
```suggestion
    public async Task<ActionResult> GetRemoteLyrics([FromRoute, Required] string lyricId)
```

More consistent with downloadRemoteLyrics and indicative of what id we want

File: Jellyfin.Api/Controllers/LyricController.cs
Comment (nielsvanvelzen) [inline Jellyfin.Api/Controllers/LyricController.cs:245]: What happens if the remote lyrics are not found or the resolver throws?

File: Jellyfin.Api/Controllers/LyricController.cs
Comment (nielsvanvelzen) [inline Jellyfin.Api/Controllers/LyricController.cs:209]: Missing 404 response type

File: Jellyfin.Api/Controllers/LyricController.cs
Comment (nielsvanvelzen) [inline Jellyfin.Api/Controllers/LyricController.cs:188]: Missing 404 response type

File: Jellyfin.Api/Controllers/LyricController.cs
Comment (nielsvanvelzen) [inline Jellyfin.Api/Controllers/LyricController.cs:77]: Missing 404 response type

File: MediaBrowser.Model/Providers/RemoteLyricInfo.cs
Comment (nielsvanvelzen) [inline MediaBrowser.Model/Providers/RemoteLyricInfo.cs:8]: I think it would be worth adding the actual lyrics as well. This can be used to "preview" the lyrics when searching lyrics manually during playback.

**Review by nielsvanvelzen — CHANGES_REQUESTED** [review]

File: Jellyfin.Api/Controllers/LyricController.cs
Comment (crobibero) [reply Jellyfin.Api/Controllers/LyricController.cs:245]: If a provider errors it won't be added to the list
I'm not sure what you mean by resolver here?

Comment (nielsvanvelzen) [reply Jellyfin.Api/Controllers/LyricController.cs]: Sorry I meant provider. When I call this endpoint with some random id that doesn't exist, what happens? I didn't see any checking here to respond with a 404.

File: Jellyfin.Api/Controllers/LyricController.cs
Comment (Bond-009) [inline Jellyfin.Api/Controllers/LyricController.cs:131]:
```suggestion
        var bytes = Encoding.UTF8.GetBytes(body.Data);
        var stream = new MemoryStream(bytes, 0, bytes.Length, false, true);
```
We can expose the internal buffer for this MemoryStream, may speed something up down the line

File: Jellyfin.Api/Controllers/LyricController.cs
Comment (crobibero) [reply Jellyfin.Api/Controllers/LyricController.cs]: Good call- this endpoint will now respond with a 404 if the provider isn't found

File: Jellyfin.Api/Controllers/SubtitleController.cs
Comment (Bond-009) [inline Jellyfin.Api/Controllers/SubtitleController.cs:413]: https://learn.microsoft.com/en-us/dotnet/standard/garbage-collection/implementing-disposeasync#unacceptable-pattern

**Review by Bond-009 — CHANGES_REQUESTED** [review]

File: Jellyfin.Api/Controllers/LyricController.cs
Comment (nielsvanvelzen) [inline Jellyfin.Api/Controllers/LyricController.cs:118]: Should be 200 now

File: MediaBrowser.Model/Lyrics/UploadLyricDto.cs
Comment (nielsvanvelzen) [inline MediaBrowser.Model/Lyrics/UploadLyricDto.cs:21]: I think "content" might be more descriptive.

File: MediaBrowser.Model/Lyrics/LyricModel.cs
Comment (nielsvanvelzen) [inline MediaBrowser.Model/Lyrics/LyricModel.cs:8]:
```suggestion
public class LyricDto
```

This would make sense for the API part but might not make much sense for the other usages

File: Jellyfin.Api/Controllers/LyricController.cs
Comment (nielsvanvelzen) [inline Jellyfin.Api/Controllers/LyricController.cs:121]:
```suggestion
    public async Task<ActionResult<LyricModel>> UploadLyrics(
```

Consistency with other endpoints

File: MediaBrowser.Model/Lyrics/RemoteLyricInfoDto.cs
Comment (nielsvanvelzen) [inline MediaBrowser.Model/Lyrics/RemoteLyricInfoDto.cs:21]: Metadata is also included in LyricModel so I guess we can drop it here

**Review by nielsvanvelzen — CHANGES_REQUESTED** [review]

File: MediaBrowser.Model/Configuration/LibraryOptions.cs
Comment (nielsvanvelzen) [inline MediaBrowser.Model/Configuration/LibraryOptions.cs:96]: Add default value to OpenAPI spec

**Review by nielsvanvelzen — APPROVED** [review]
✔️ API changes

[Kotlin SDK diff](https://github.com/jellyfin/jellyfin-sdk-kotlin/compare/08c55eb27d45e195a4a229a6abb012628014bf1a...nielsvanvelzen:jellyfin-sdk-kotlin:tst-lyrics)

File: MediaBrowser.Providers/Lyric/LyricManager.cs
Comment (crobibero) [reply MediaBrowser.Providers/Lyric/LyricManager.cs:143]: Can't use `ReadOnlySpan` in async methods

I can create some string versions of `LeftPart` and `RightPart` if desired

Comment (Bond-009) [reply MediaBrowser.Providers/Lyric/LyricManager.cs:143]: You can use it, you just can't have a ref struct variable in an async method
```suggestion
        var provider = GetProvider(lyricId.AsSpan().LeftPart('_').ToString());
```
This should work (if you add `using Jellyfin.Extensions;`)

string versions of `LeftPart` and `RightPart` would make this a bit cleaner, but I'm not sure if that's needed

File: MediaBrowser.Providers/MediaInfo/MediaInfoResolver.cs
Comment (Bond-009) [inline MediaBrowser.Providers/MediaInfo/MediaInfoResolver.cs:181]: Can't this be an array?

File: MediaBrowser.Providers/MediaInfo/MediaInfoResolver.cs
Comment (Bond-009) [inline MediaBrowser.Providers/MediaInfo/MediaInfoResolver.cs:194]: Normally we don't use `AsReadOnly()` bc every time you use it afterwards everything gets proxied trough `ReadOnlyCollection` which slows everything down

File: MediaBrowser.Providers/Lyric/LyricManager.cs
Comment (Bond-009) [inline MediaBrowser.Providers/Lyric/LyricManager.cs:98]: Maybe this `ToList` should happen outside of the `OrderBy`?

**Review by Bond-009 — CHANGES_REQUESTED** [review]

File: MediaBrowser.Providers/Lyric/LyricManager.cs
Comment (crobibero) [reply MediaBrowser.Providers/Lyric/LyricManager.cs]: Turns out I had already added a `.IndexOf` for `IReadOnlyList`

That's what I get from just pulling code from the SubtitleManager

**Review by Bond-009 — APPROVED** [review]

**Comment by nielsvanvelzen** [comment]
I will need some extra time to review with the latest changes because the Kotlin SDK does not currently support multipart/form-data so I need to implement that first.

**Review by nielsvanvelzen — CHANGES_REQUESTED** [review]
The schema for the requestBody of the upload endpoint is now generated inline which is impossible to work with in the Kotlin SDK.

**Review by nielsvanvelzen — APPROVED** [review]
✔️ API changes

[Kotlin SDK diff](https://github.com/jellyfin/jellyfin-sdk-kotlin/compare/08c55eb27d45e195a4a229a6abb012628014bf1a...nielsvanvelzen:jellyfin-sdk-kotlin:tst-lyrics)

**Comment by DomiStyle** [comment]
I would still like to add support for embedded lyrics but I'm unsure what the preferred way to achieve this with the new structure is.

Some things I thought of:

* Add a new `MediaStreamType`: `EmbeddedLyrics`
* Add a `Lyrics` or `EmbeddedLyrics` attribute to the `Audio` class - probably problematic since all the lyrics are assumed to be a `MediaStream` now
* Add the embedded lyrics as text directly as `MediaStreamType.Lyric` - would need special handling to separate lyrics that are paths vs lyrics that are text
* Add the path of the audio file as `MediaStreamType.Lyric` directly instead of to the `.lrc` file and get the embedded lyrics when lyrics are requested via the API - probably not ideal since the tags have to be read every time the lyrics are requested

Any preferred way to do this before I start work on a PR?

**Comment by NicholasFlamy** [comment]
> I would still like to add support for embedded lyrics but I'm unsure what the preferred way to achieve this with the new structure is.
> 
> Some things I thought of:
> 
> * Add a new `MediaStreamType`: `EmbeddedLyrics`
> * Add a `Lyrics` or `EmbeddedLyrics` attribute to the `Audio` class - probably problematic since all the lyrics are assumed to be a `MediaStream` now
> * Add the embedded lyrics as text directly as `MediaStreamType.Lyric` - would need special handling to separate lyrics that are paths vs lyrics that are text
> * Add the path of the audio file as `MediaStreamType.Lyric` directly instead of to the `.lrc` file and get the embedded lyrics when lyrics are requested via the API - probably not ideal since the tags have to be read every time the lyrics are requested
> 
> Any preferred way to do this before I start work on a PR?

I seem to remember reading that the embedded metadata handling was going to be refactored or something.

**Comment by crobibero** [comment]
@DomiStyle @NicholasFlamy please test out #11081

I only have audio files that I manually embedded lyrics into .mp3 and .flac using MediaMonkey5

**Comment by NicholasFlamy** [comment]
> please test out #11081

I am not familiar with how to download the build artifacts from a Jellyfin PR.

**Comment by NicholasFlamy** [comment]
Either this or #11081 might be causing Media Info to be blank.
Screenshot 2024-03-08 113309
It only happens to songs with lyrics, embedded or LRC file.

I just tested this. Some songs with embedded lyrics won't show Media Info, they look like that screenshot above. Then I tested a song that was encoded with the exact same encoder and settings and the same resolution (1200x1200) mozjpeg-encoded embedded cover art but no embedded lyrics or LRC file. Then I copied a random LRC file and moved it to the directory with this song with no lyrics and renamed the LRC file to match the song (but with the .lrc extension rather than .m4a of course) and rescanned and refreshed Jellyfin and the song had the same problem as the one with embedded lyrics.

My encodes are using these FFmpeg args `-c:a libfdk_aac -vbr 4 -cutoff 20000 -ar 44100 -movflags +faststart` in case it's specific to this.

I'm running the unstable docker image in TrueNAS Scale. I did not test this before these two issues were merged so I cannot confirm that these caused it.

Screenshot of song without lyrics (no embedded lyrics and no .lrc file with the same name as the file)
image

**Comment by crobibero** [comment]
@NicholasFlamy please create an issue will full steps so it can be tracked appropriately

**Comment by NicholasFlamy** [comment]
> @NicholasFlamy please create an issue will full steps so it can be tracked appropriately

#11130 Gotcha

**Pull request merged**
